### PR TITLE
Extending pipeline promotions token security documentation 

### PR DESCRIPTION
**Pull request by enekofb**
Closes https://github.com/weaveworks/weave-gitops-enterprise/issues/2564

**What changed?**

Extending existing guidance to include the ability to create deny semantics required for least privilege on access to the credentials token

**Why was this change made?**
Cause otherwise, pipeline controller would have access to the secret but nothing would have avoid that other roles has access to.

**How was this change implemented?**
Docs

**How did you validate the change?**

Tested locally 

Screenshot 2023-07-11 at 09 54 14

**Release notes**

**Documentation Changes**

### CODE REVIEW

File: website/docs/pipelines/promoting-applications.mdx
Comment (enekofb) [inline website/docs/pipelines/promoting-applications.mdx:299]: Screenshot 2023-03-28 at 09 27 41

- policy or policy config https://docs.gitops.weave.works/docs/0.18.0/policy/policy-configuration/

File: website/docs/pipelines/promoting-applications.mdx
Comment (bigkevmcd) [inline website/docs/pipelines/promoting-applications.mdx:317]: Can we do a couple of things here?
 1. Put the full version of this for users so that they can copy and paste.
 2. Check what happens when someone uses `*` for verbs

**Comment by enekofb** [comment]
@enekofb @squaremo 

Assumptions

Extend RBAC 
- clusterrolebindings not allowed only admins
- rolebindings not allowed for non-pipeline namespace

Policy
- restrict wildcards
- restrict list 
- only allow get secret resource by name to pipeline-controller SA

File: website/docs/pipelines/promoting-applications.mdx
Comment (enekofb) [inline website/docs/pipelines/promoting-applications.mdx:165]: Extend with ours too

File: website/docs/pipelines/promoting-applications.mdx
Comment (enekofb) [inline website/docs/pipelines/promoting-applications.mdx:144]: Extract the general recommendations to a general-purpose page and reference it here. it is required for other cases.

File: website/docs/pipelines/promoting-applications.mdx
Comment (enekofb) [inline website/docs/pipelines/promoting-applications.mdx:153]: >lower environments 
low risk environment

File: website/docs/pipelines/promoting-applications.mdx
Comment (enekofb) [inline website/docs/pipelines/promoting-applications.mdx:161]: extend with questions that you should ask as platform admin to make it a workable scenario - for example shared configuration repo (test, staging, prod)

we cannot provide checklist but guidance

ex solution leverage codeowners for extra eyes for production  

to review in the context of the shared responsibility model

File: website/docs/pipelines/promoting-applications.mdx
Comment (enekofb) [inline website/docs/pipelines/promoting-applications.mdx:176]: do not extract get of the equation

Comment (enekofb) [reply website/docs/pipelines/promoting-applications.mdx:176]: what about the scenario where granting get access to a service account and someone init he namespace can run pods with that service account

File: website/docs/pipelines/promoting-applications.mdx
Comment (enekofb) [inline website/docs/pipelines/promoting-applications.mdx:184]: provide running example

Comment (enekofb) [reply website/docs/pipelines/promoting-applications.mdx:184]: what about templates

File: website/docs/pipelines/promoting-applications.mdx
Comment (enekofb) [inline website/docs/pipelines/promoting-applications.mdx:252]: extend 

1) be careful with rolebindings out of the namespace - for example, allowing, extending 
2) dont give kubectl access with rbac write access because as a user i could escape this control

File: website/docs/pipelines/promoting-applications.mdx
Comment (enekofb) [inline website/docs/pipelines/promoting-applications.mdx:295]: controller should live in its namespace

File: website/docs/pipelines/promoting-applications.mdx
Comment (enekofb) [inline website/docs/pipelines/promoting-applications.mdx:379]: extend with following guidance for git provider and need to update the secret after it expires.

**Review by enekofb — COMMENTED** [review]
comments added from session with Michael and Kevin

File: website/docs/pipelines/promoting-applications.mdx
Comment (enekofb) [reply website/docs/pipelines/promoting-applications.mdx:150]: done

File: website/docs/pipelines/promoting-applications.mdx
Comment (enekofb) [reply website/docs/pipelines/promoting-applications.mdx:153]: updated

File: website/docs/pipelines/promoting-applications.mdx
Comment (enekofb) [reply website/docs/pipelines/promoting-applications.mdx:161]: updated

File: website/docs/pipelines/promoting-applications.mdx
Comment (enekofb) [reply website/docs/pipelines/promoting-applications.mdx]: >what about the scenario where granting get access to a service account and someone init he namespace can run pods with that service account
added

File: website/docs/pipelines/promoting-applications.mdx
Comment (enekofb) [reply website/docs/pipelines/promoting-applications.mdx:237]: >dont give kubectl access with rbac write access because as a user i could escape this control

added

File: website/docs/pipelines/promoting-applications.mdx
Comment (enekofb) [reply website/docs/pipelines/promoting-applications.mdx]: adding https://github.com/weaveworks/weave-gitops-quickstart/pull/5

File: website/docs/pipelines/promoting-applications.mdx
Comment (enekofb) [inline website/docs/pipelines/promoting-applications.mdx:209]: @MostafaMegahid how should we contribute these policies to the policy library ?

File: website/docs/pipelines/promoting-applications.mdx
Comment (enekofb) [reply website/docs/pipelines/promoting-applications.mdx]: Given policy library is enterprise feature, I cannot add it here.

File: website/docs/pipelines/promoting-applications.mdx
Comment (enekofb) [reply website/docs/pipelines/promoting-applications.mdx]: updated

File: website/docs/pipelines/promoting-applications.mdx
Comment (enekofb) [reply website/docs/pipelines/promoting-applications.mdx]: Today we synced on this topic. the suggested path would be to 

1. add these policies to [policy library](https://github.com/weaveworks/policy-library/tree/main/policies). An example of this could https://github.com/weaveworks/policy-library/tree/main/policies/RBACClusterRoleClusterAdmin

Comment (enekofb) [reply website/docs/pipelines/promoting-applications.mdx]: Draft PR for good practices within policy-library https://github.com/weaveworks/policy-library/pull/31

File: website/docs/pipelines/promoting-applications.mdx
Comment (enekofb) [inline website/docs/pipelines/promoting-applications.mdx:208]: review with @MostafaMegahid whether this is the best way to use it or we should use other sources

File: website/docs/pipelines/promoting-applications.mdx
Comment (squaremo) [inline website/docs/pipelines/promoting-applications.mdx:393]:
```suggestion
    - Tokens close to their expiration date are cycled.
```

File: website/docs/pipelines/promoting-applications.mdx
Comment (squaremo) [inline website/docs/pipelines/promoting-applications.mdx:455]:
```suggestion
1. Pipeline controller receives events via [webhook from leaf clusters](./#setup-notifications-from-leaf-clusters). An [HMAC](https://en.wikipedia.org/wiki/HMAC) is used for verification so a shared key should be provided in this case.
```

File: website/docs/pipelines/promoting-applications.mdx
Comment (squaremo) [inline website/docs/pipelines/promoting-applications.mdx:429]: I think from here down it's just moved text, is that right? (feel free to ignore comments below here!)

File: website/docs/pipelines/promoting-applications.mdx
Comment (squaremo) [inline website/docs/pipelines/promoting-applications.mdx:162]:
```suggestion
- In the case of [repo per environment](https://fluxcd.io/flux/guides/repository-structure/#repo-per-environment) you want to ensure that a pipeline is configured with your test environment repo URL. You also want to ensure that you have segregation of tokens per environment rather than allowing a token to access any environment repo.
```

File: website/docs/pipelines/promoting-applications.mdx
Comment (squaremo) [inline website/docs/pipelines/promoting-applications.mdx:196]:
```suggestion
good practices policies that will help you keep pipeline secrets secure according to the previous RBAC recommendations. Deploy
```

**Review by squaremo — COMMENTED** [review]
Just edits as suggestions, this time. Once we have nailed down how to use the policy library, I think we're good to go.

File: website/docs/pipelines/promoting-applications.mdx
Comment (enekofb) [reply website/docs/pipelines/promoting-applications.mdx:154]: We will refactor after merged so we dont increase scope from the original one.

File: website/docs/pipelines/promoting-applications.mdx
Comment (enekofb) [reply website/docs/pipelines/promoting-applications.mdx:297]: This concern it is reflected in the shared security context board. it will be part of the future security documentation.

File: website/docs/pipelines/promoting-applications.mdx
Comment (enekofb) [reply website/docs/pipelines/promoting-applications.mdx]: To review today with Mostafa

File: website/docs/pipelines/promoting-applications.mdx
Comment (enekofb) [reply website/docs/pipelines/promoting-applications.mdx:208]: To review today with Mostafa.

File: website/docs/pipelines/promoting-applications.mdx
Comment (enekofb) [reply website/docs/pipelines/promoting-applications.mdx:237]: This example shows how you could leverage PolicyConfig to add exclusions for privilege workloads for your policies. 

In our example, we are deploying the goodpractices kustomization that deploy policies like `weave.templates.rbac-prohibit-wildcards-policyrule-resources`. These policies by default does not have exclusions. In case you want to have exclusions (see my comment [here](https://github.com/weaveworks/policy-library/pull/31/files#r1273143158)) you leverage PolicyConfig.

File: website/docs/pipelines/promoting-applications.mdx
Comment (enekofb) [reply website/docs/pipelines/promoting-applications.mdx:264]: it tries to call out that if you have exclusions, they could be used as an escalation mechanism. Given exclusions appears in the doc first time here, we just try to call out / reinforce the risk and how to mitigate it.

File: website/docs/pipelines/promoting-applications.mdx
Comment (enekofb) [reply website/docs/pipelines/promoting-applications.mdx:429]: Umms, I could not see comments for this section: feel free to add them to review them too. Thanks!

File: website/docs/pipelines/promoting-applications.mdx
Comment (enekofb) [reply website/docs/pipelines/promoting-applications.mdx:208]: From the conversation with Mostafa: 

- We expect policy library to be used as a set of examples that customers might need to customize.
- The way we expect now to do so is via Git, copy and paste then to update with the values that better suit them.
- Library is enterprise and its value would be protected by the ww - customer agreement.

Comment (squaremo) [reply website/docs/pipelines/promoting-applications.mdx:208]: How do people get access to it?

File: website/docs/pipelines/promoting-applications.mdx
Comment (squaremo) [reply website/docs/pipelines/promoting-applications.mdx:429]: There was just one I think; I didn't bother reviewing most of the text that you didn't change.

File: website/docs/pipelines/promoting-applications.mdx
Comment (enekofb) [reply website/docs/pipelines/promoting-applications.mdx:208]: To follow up next week in policy distribution out of git

File: website/docs/pipelines/promoting-applications.mdx
Comment (enekofb) [inline website/docs/pipelines/promoting-applications.mdx:200]: provide guidance on how to leverage CREs for helping you

File: website/docs/pipelines/promoting-applications.mdx
Comment (enekofb) [inline website/docs/pipelines/promoting-applications.mdx:201]: @squaremo let me know your thoughts

Screenshot 2023-07-26 at 11 46 02

File: website/docs/pipelines/promoting-applications.mdx
Comment (enekofb) [reply website/docs/pipelines/promoting-applications.mdx:205]: updated https://github.com/weaveworks/weave-gitops/pull/3540/files#r1274777480

File: website/docs/pipelines/promoting-applications.mdx
Comment (squaremo) [reply website/docs/pipelines/promoting-applications.mdx:201]: If those are real job titles, I'm on board with it :-)

**Review by squaremo — APPROVED** [review]
Thorough work Eneko, thank you so much for following this through. What a journey :-)

**Pull request merged**
